domains/pallets/messenger/src/lib.rs

```diff
@@ -1042,23 +1057,46 @@ mod pallet {
             // verify and decode message
             let msg = Self::do_verify_xdm(next_nonce, key, xdm)?;
 
-            // if there is no channel config, this must be the Channel open request
-            if should_init_channel {
-                match msg.payload {
-                    VersionedPayload::V0(Payload::Protocol(RequestResponse::Request(
-                        ProtocolMessageRequest::ChannelOpen(_),
-                    ))) => {}
-                    _ => {
-                        log::error!("Unexpected call instead of channel open request: {:?}", msg,);
-                        return Err(InvalidTransaction::Call.into());
+            let is_valid_call = match &msg.payload {
+                VersionedPayload::V0(payload) => match payload {
+                    Payload::Protocol(RequestResponse::Request(req)) => match req {
+                        // channel open should ensure there is no Channel present already
+                        ProtocolMessageRequest::ChannelOpen(_) => maybe_channel.is_none(),
+                        // we allow channel close only if it is init or open state
+                        ProtocolMessageRequest::ChannelClose => {
+                            if let Some(ref channel) = maybe_channel {
+                                !(channel.state == ChannelState::Closed)
+                            } else {
+                                false
+                            }
+                        }
+                    },
+                    // endpoint request messages are only allowed when
+                    // channel is open, or
+                    // channel is closed. Channel can be closed by dst_chain simultaneously
+                    // while src_chain already sent a message. We allow the message but return an
+                    // error in the response so that src_chain can revert any necessary actions
+                    Payload::Endpoint(RequestResponse::Request(_)) => {
+                        if let Some(ref channel) = maybe_channel {
+                            !(channel.state == ChannelState::Initiated)
```

Comment:
This is only valid when src_chain has not closed the channel yet but dst_chain did. This is the expected change to ensure in transit messages are processed.

Once the channel is closed, src_chain should not initiate new messages

Reply:
If the src domain is compromised i.e. controlled by a single entity then they can filter the close channel response when constructing bundle to keep the channel open on the src domain and keep sending XDM to the dst domain.

Reply:
This problem does not exist if the src_domain is sending messages since it wont be able to send any messages after the channel is closed. Any messages sent before channel is closed will be processed and responses are executed.

If the dst_chain closes the channel they will not be able to send any messages to src_chain but sure the operator of the src_chain may exclude the Channel close message from the dst_chain in the bundlee. Any messages sent from src_chain to dst_chain, since src_chain did not receive the message to close channel,  will be responded with Err and necessary fee is taken by the relayer. So there is a cost associated with this.

The reason for this change it to ensure following
dst_chain closes the channel but simultaneouly src_Chain sends messages. These messages wont be accepted and relayer will continue to try and fail. Allowing messages to land then would unblock the relayer while they still get their fee.

IMHO, while they can send messages and pay for the relay fee, this will not harm the dst_chain since it always reponds Err.


If you have a better approach that can solve the audit issue mentioned above, please mention them